### PR TITLE
Label display issues were fixed

### PR DESCRIPTION
**Pull request by zBritva**
Label text size issue was fixed
Label display issue was fixed
Choose of axis orientation beginning was added
Min value at zero options was added
Sorting capabilities were added

### CODE REVIEW

File: src/radarChart.ts [hunk 1]
```diff
@@ -471,7 +506,7 @@ module powerbi.extensibility.visual {
                 options.element);
 
             this.legend = createLegend(
-                $(element),
+                jQuery(element),
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Why do you use ```jQuery``` instead of ```$```?

File: src/radarChart.ts [hunk 2]
```diff
@@ -607,6 +647,7 @@ module powerbi.extensibility.visual {
         }
 
         private drawCircularSegments(values: PrimitiveValue[]): void {
+            let axisBeginning = this.radarChartData.settings.axisBeginning;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 3]
```diff
@@ -650,6 +691,7 @@ module powerbi.extensibility.visual {
         }
 
         private drawAxes(values: PrimitiveValue[]): void {
+            let axisBeginning = this.radarChartData.settings.axisBeginning;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 4]
```diff
@@ -702,6 +744,125 @@ module powerbi.extensibility.visual {
...
+                return true;
+            }
+            if (y22 <= y12 && y12 <= y21) {
+                return true;
+            }
+            return false;
+        }
+
+        private shiftText(currentTextY1: number, currentTextY12: number, otherTextY21: number, otherTexty22: number, direction: boolean): number {
+            let shift = 0;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 5]
```diff
@@ -702,6 +744,125 @@ module powerbi.extensibility.visual {
...
+        private shiftIntersectText(current: RadarChartLabel, others: RadarChartLabel[], shiftDown: boolean): void {
+            let labelSettings: RadarChartLabelSettings = this.radarChartData.settings.labels;
+
+            let properties: TextProperties = {
+                fontFamily: RadarChart.AxesLabelsFontFamily,
+                fontSize: PixelConverter.fromPoint(labelSettings.fontSize),
+                text: this.radarChartData.labels.formatter.format(current.text)
+            };
+
+            let currentTextHeight = textMeasurementService.estimateSvgTextHeight(properties);
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 6]
```diff
@@ -702,6 +744,125 @@ module powerbi.extensibility.visual {
...
+
+            for (let i: number = 0; i < others.length; i++) {
+                let label: RadarChartLabel = others[i];
+
+                properties.text = label.text;
+                let otherTextHeight = textMeasurementService.estimateSvgTextHeight(properties);
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 7]
```diff
@@ -702,6 +744,125 @@ module powerbi.extensibility.visual {
...
+
+                let curTextUpperPoint = current.y - currentTextHeight;
+                let labelTextUpperPoint = label.y - otherTextHeight;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 8]
```diff
@@ -702,6 +744,125 @@ module powerbi.extensibility.visual {
...
+
+            let currentTextHeight = textMeasurementService.estimateSvgTextHeight(properties);
+
+            for (let i: number = 0; i < others.length; i++) {
+                let label: RadarChartLabel = others[i];
+
+                properties.text = label.text;
+                let otherTextHeight = textMeasurementService.estimateSvgTextHeight(properties);
+
+                let curTextUpperPoint = current.y - currentTextHeight;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 9]
```diff
@@ -702,6 +744,125 @@ module powerbi.extensibility.visual {
...
+                let labelTextUpperPoint = label.y - otherTextHeight;
+
+                if (this.isIntersect(current.y, curTextUpperPoint, label.y, labelTextUpperPoint)) {
+                    let shift = this.shiftText(current.y, curTextUpperPoint, label.y, labelTextUpperPoint, shiftDown);
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 10]
```diff
@@ -702,6 +744,125 @@ module powerbi.extensibility.visual {
...
+                    if (!shiftDown && current.y < 0 || shiftDown && current.y + currentTextHeight / 2 > 0) {
+                        current.hide = true;
+                    }
+                }
+            }
+        }
+
+        private shiftCollidedLabels(labelPoints: RadarChartLabel[]): void {
+            // from 0 to 90 shift up by Y
+            let maxRadius = 0;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 11]
```diff
@@ -702,6 +744,125 @@ module powerbi.extensibility.visual {
...
+            labelPoints.forEach(point => {
+                point.x > maxRadius ? maxRadius = point.x : 0;
+            });
+
+            let shiftDirrectionIsDown = this.radarChartData.settings.axisBeginning === 1;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 12]
```diff
@@ -713,18 +874,38 @@ module powerbi.extensibility.visual {
                 radius: number = this.radius,
                 labelPoints: RadarChartLabel[] = this.radarChartData.labels.labelPoints;
 
+            let axisBeginning = this.radarChartData.settings.axisBeginning;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 13]
```diff
@@ -713,18 +874,38 @@ module powerbi.extensibility.visual {
...
+
+                label.xLinkBegin = radius * RadarChart.LabelLinkBeginPositionFactor * Math.sin(angleInRadian);
+                label.yLinkBegin = axisBeginning * radius * RadarChart.LabelLinkBeginPositionFactor * Math.cos(angleInRadian);
 
                 label.textAnchor = (i * angle) < Math.PI
                     ? RadarChart.TextAnchorStart
                     : RadarChart.TextAnchorEnd;
+            }
+
+            let maxRadius = 0;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 14]
```diff
@@ -713,18 +874,38 @@ module powerbi.extensibility.visual {
...
+            labelPoints.forEach(point => {
+                point.x > maxRadius ? maxRadius = point.x : 0;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Should we bind zero to ```maxRadius``` if ```point.x < maxRadius```?

File: src/radarChart.ts [hunk 15]
```diff
@@ -734,25 +915,26 @@ module powerbi.extensibility.visual {
...
             // Hide and reposition labels that overlap
-            let dataLabelManager: DataLabelManager = new DataLabelManager(),
-                filteredData: LabelEnabledDataPoint[] = dataLabelManager.hideCollidedLabels(
-                    this.viewport,
-                    labelPoints,
-                    labelLayout,
-                    true);
-
-            this.drawAxesLabels(filteredData as RadarChartLabel[]);
+            // let dataLabelManager: DataLabelManager = new DataLabelManager(),
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please get rid of commented code lines.

File: src/radarChart.ts [hunk 16]
```diff
@@ -734,25 +915,26 @@ module powerbi.extensibility.visual {
...
             let labelSettings: RadarChartLabelSettings = this.radarChartData.settings.labels;
 
-            let selection: d3.Selection<RadarChartLabel> = this.mainGroupElement
+            let selectionLableText: d3.Selection<RadarChartLabel> = this.mainGroupElement
                 .select(RadarChart.AxisSelector.selector)
                 .selectAll(RadarChart.AxisLabelSelector.selector);
 
-            let labelsSelection: UpdateSelection<RadarChartLabel> = selection.data(
-                values.filter((label: RadarChartLabel) => labelSettings.show));
+            let filteredData = values.filter((label: RadarChartLabel) => labelSettings.show && !label.hide);
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 17]
```diff
@@ -764,7 +946,10 @@ module powerbi.extensibility.visual {
                     transform: translate(
                         RadarChart.LabelXOffset,
                         -RadarChart.LabelYOffset * labelSettings.fontSize),
-                    x: (label: RadarChartLabel) => label.x,
+                    x: (label: RadarChartLabel) => {
+                        let shift = label.textAnchor === RadarChart.TextAnchorStart ? +RadarChart.LabelPositionXOffset : -RadarChart.LabelPositionXOffset;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 18]
```diff
@@ -784,18 +969,68 @@ module powerbi.extensibility.visual {
...
+            labelsShortLineLinkSelection
+                .enter()
+                .append("svg:line");
+
+            labelsShortLineLinkSelection
+                .attr({
+                    x1: (label: RadarChartLabel) => label.xLinkEnd,
+                    y1: (label: RadarChartLabel) => label.yLinkEnd,
+                    x2: (label: RadarChartLabel) => {
+                        let shift = label.textAnchor === RadarChart.TextAnchorStart ? +(RadarChart.LabelPositionXOffset - 2) : -(RadarChart.LabelPositionXOffset - 2);
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 19]
```diff
@@ -784,18 +969,68 @@ module powerbi.extensibility.visual {
...
+                .exit()
+                .remove();
         }
 
         private drawChart(series: RadarChartSeries[], duration: number): void {
             let angle: number = this.angle,
                 dataPoints: RadarChartDatapoint[][] = this.getDataPoints(series),
                 layers: RadarChartDatapoint[][] = d3.layout.stack<RadarChartDatapoint>()(dataPoints),
                 yDomain: any = this.calculateChartDomain(series);
+            let axisBeginning = this.radarChartData.settings.axisBeginning;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: src/radarChart.ts [hunk 20]
```diff
@@ -930,24 +1165,16 @@ module powerbi.extensibility.visual {
...
 
-            let minValue: number = d3.min(dataPointsList, (dataPoint: RadarChartDatapoint) => {
-                return dataPoint.y;
-            });
-
             let maxValue: number = d3.max(dataPointsList, (dataPoint: RadarChartDatapoint) => {
                 return dataPoint.y;
             });
 
+            let minValue = this.radarChartData.settings.minValue;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please specify a type of the variable explicitly.

File: package.json [hunk 21]
```diff
@@ -14,13 +14,19 @@
...
+    "test": "karma start",
+    "debug": "karma start --browsers=Chrome --single-run=false"
   },
   "license": "MIT",
   "repository": {
     "type": "git",
     "url": "git+https://github.com/Microsoft/PowerBI-visuals-RadarChart.git"
   },
+  "dependencies": {
+    "jscommon": "0.1.0",
```
Comment (ignatvilesov) [inline package.json]: Does anything use it?

File: src/radarChart.ts (hunk 14 above)
Comment (zBritva) [inline src/radarChart.ts]: @ignatvilesov it's

File: package.json (hunk 21 above)
Comment (zBritva) [inline package.json]: @ignatvilesov no anymore. removed

**Comment by zBritva** [comment]
@ignatvilesov I fixed

File: src/radarChart.ts (hunk 14 above)
Comment (ignatvilesov) [inline src/radarChart.ts]: Please use the plain ```if``` instead. It would improve readability.

File: src/radarChart.ts [hunk 22]
```diff
@@ -713,46 +875,53 @@ module powerbi.extensibility.visual {
...
-
-                label.maxWidth = this.viewportAvailable.width - Math.abs(label.x);
             }
 
-            let labelArc: SvgArc<Arc> = d3.svg.arc()
-                .innerRadius(() => radius)
-                .outerRadius(() => radius * RadarChart.OuterRadiusFactor);
+            let maxRadius: number = 0;
+            labelPoints.forEach(point => {
+                point.x > maxRadius ? maxRadius = point.x : 0;
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please use the common ```if```` instead.

File: src/radarChart.ts [hunk 23]
```diff
@@ -784,18 +926,68 @@ module powerbi.extensibility.visual {
             labelsSelection
                 .exit()
                 .remove();
+
+            let selectionLongLIneLableLink: d3.Selection<RadarChartLabel> = this.mainGroupElement
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Please note that ```I``` in the ```selectionLongLIneLableLink``` should be lowercase.

Comment (zBritva) [inline src/radarChart.ts]: @ignatvilesov fixed

File: src/radarChart.ts [hunk 24]
```diff
@@ -677,29 +718,125 @@ module powerbi.extensibility.visual {
...
+            let maxRadius: number = 0;
+            labelPoints.forEach(point => {
+                if (Math.abs(point.x) > maxRadius) {
+                    maxRadius = Math.abs(point.x);
+                }
+            });
+
+            let shiftDirrectionIsDown: boolean = this.radarChartData.settings.axisBeginning === 1;
+
+            for (let i: number = 0; i < labelPoints.length; i++) {
```
Comment (AviSander) [inline src/radarChart.ts]: consider refactoring the 2 for() loops below. There's some code repetition which can be refactored out for better readability.

File: src/radarChart.ts [hunk 25]
```diff
@@ -1088,7 +1302,7 @@ module powerbi.extensibility.visual {
          * This function returns the values to be displayed in the property pane for each object.
          * Usually it is a bind pass of what the property pane gave you, but sometimes you may want to do
          * validation and return other values/defaults
-         * 
+         *
          * TODO: We should use SettingsParser instead. Please rewrite it in future versions.
          */
         public enumerateObjectInstances(options: EnumerateVisualObjectInstancesOptions): VisualObjectInstanceEnumeration {
```
Comment (ignatvilesov) [inline src/radarChart.ts]: We should use `DataViewObjectParser` to simplify it. Let's include it to our backlog.

File: test/visualTest.ts [hunk 26]
```diff
@@ -331,6 +331,24 @@ module powerbi.extensibility.visual.test {
                         });
                 });
             });
+
+            describe("In visual with small size", () => {
```
Comment (ignatvilesov) [inline test/visualTest.ts]: Please rename this test to provide more readability. For instance: `Data labels at small viewport`.

File: src/radarChart.ts [hunk 27]
```diff
@@ -1041,6 +1228,28 @@ module powerbi.extensibility.visual {
                 objects = dataView.metadata.objects;
             }
 
+            let min: number = <number>dataView.categorical.values[0].values[0];
+            for (let values in dataView.categorical.values) {
```
Comment (ignatvilesov) [inline src/radarChart.ts]: Can we use _.min or d3.min?

File: src/radarChart.ts (hunk 24 above)
Comment (zBritva) [inline src/radarChart.ts]: @AviSander I did

**Pull request merged**